src/NHibernate/Engine/TwoPhaseLoad.cs

```diff
@@ -84,15 +85,28 @@ public static void InitializeEntity(object entity, bool readOnly, ISessionImplem
 				log.Debug("resolving associations for {0}", MessageHelper.InfoString(persister, id, session.Factory));
 
 			IType[] types = persister.PropertyTypes;
+			var collectionToResolveIndexes = new List<int>();
```

Comment:
Probably give it an initial capacity to avoid resizing 

```suggestion
			var collectionToResolveIndexes = new List<int>(hydratedState.Length);
```